Spellchecker

DronKonstantin opened the pull request:
Исправил код в соответствии со сделанными замечаниями

DronKonstantin committed:
Created Interfaces and SpellCheckerLT

Создан модуль проверки правописания:
1. Описаны интерфейсы для взаимодействия с API или стронней библиотекой
проверки правописания
2. Реализован интерфейс в виде класса взаимодействия с API LanguageTool
3. Созданы синхронные и асинхронные методы для получения списка
доступных языков и проверки текста
4. В классе Main реализованы два метода для демонстрации работы
синхронных и асинхронных методов

DronKonstantin committed:
Доработка

1. Немного переработан код, добавлена обработка ошибок.
2. Добавлены подробные комментарии

Climber24 commented on BOTFirst/SpellChecker/Interfaces/ISpellChecker.cs:
```diff
@@ -0,0 +1,24 @@
+{
+    public delegate void GetAvailableLanguagesAsyncEventHandler(); // делегаты созданные для событий асинхронных методов
+    public delegate void CheckAsyncContainerEventHandler();
+    interface ISpellChecker
```
Может быть стоит этот интерфейс тоже сделать абстрактным классом?
Из code convention'ов:
✓ DO name interfaces with adjective phrases, or occasionally with nouns or noun phrases.
Nouns and noun phrases should be used rarely and they might indicate that the type should be an abstract class, and not an interface.

parshikovalexey replied:
@DronKonstantin, @LaHesis. Объясню почему просил сделать это на интерфейсах. У интерфейсов и абстрактных классов есть общие черты, когда они оба описывают сигнатуры методов, которые должны быть реализованы в наследниках. Но при этом в абстрактном классе в самом можно сразу реализовать какой-то функционал, который будет использоваться в наследниках. Интерфейсы же только описывают то, что должно быть реализовано в наследниках. Второе большое различие. Класс может наследовать множество интерфейсов и реализовывать их, но может наследовать только один абстрактный класс.
Теперь к нашим задачам. Я предлагаю использовать интерфейсы для того, чтобы не использовать возможность чего-то там зашить в функционал самого асбстрактного класса, то есть сделать само описание интерфейса прозрачным и не опасаться появления внутри него каких-либо концов для возможных проблем. Во-вторых, как я покажу позднее, в проектах web-api есть возможность задать при формировании контекста выполнения приложения - определенные реализации интерфейса для описанных в проекте интерфейсов. И не нужно будет нигде в коде в конструкторах создавать экземпляры объектов. Для абстрактных классов - так сделать не получится.

Climber24 commented on BOTFirst/SpellChecker/Classes/Mistake.cs:
```diff
@@ -0,0 +1,24 @@
+        public int Length;
+        public Position (int begin, int len)
+        {
+            Begin = begin;
```
Полагаю, что следует добавить this в конструкторе; понятно, что по идее можно не добавлять, но об этом пишут в code conventions, что это улучшает читабельность.

parshikovalexey replied:
Согласен, будет читабельнее.

DronKonstantin replied:
Добавлю

Climber24 commented on BOTFirst/SpellChecker/Classes/SpellCheckerLT.cs:
```diff
@@ -0,0 +1,247 @@
+namespace BOTFirst.Spellchecker
+{
+
+    public class SpellCheckerLT : ISpellChecker
```
Может быть SpellCheckerLT стоит сделать статичным классом?

parshikovalexey replied:
@LaHesis, этот класс реализует интерфейс, а при использовании интерфейса - нам в любом случае придется делать явное/неявное создание экземпляра, который его будет реализовывать (то есть Static нецелесообразно)

Climber24 commented on BOTFirst/SpellChecker/Classes/SpellCheckerLT.cs:
```diff
@@ -0,0 +1,247 @@
+        #region Приватные поля
+        private List<Mistake> mistakes; // дублирующее поле содержащее список ошибок в проверенном тексте
+        private Dictionary<string, string> languages; // словарь доступных для проверки языков где ключ-название языка(на английском), значение - код языка
+        private string checkUrl; // URL для проверки текста
```
Я бы предложил сделать URL'ы readonly и переименовать checkUrl в urlForChecking, checkUrl лично у меня скорее ассоциируется с методом, чем с переменной.

Climber24 commented on BOTFirst/SpellChecker/Classes/SpellCheckerLT.cs:
```diff
@@ -0,0 +1,247 @@
+        private string checkUrl; // URL для проверки текста
+        private string getLanguagesUrl; // URL для получения списка доступных языков
+        private string checkResult; // строка(JSON) полученная от LanguageTool API при проверке текста
+        private string languagesresult; // строка(JSON) полученная от LanguageTool API при получении списка доступных языков
```
languagesResult?

Climber24 commented on BOTFirst/SpellChecker/Classes/SpellCheckerLT.cs:
```diff
@@ -0,0 +1,247 @@
+        #endregion
+    }
+    
+    public class MistakeLT : Mistake
```
Не вынести ли MistakeLT в отдельный файл?

DronKonstantin replied:
Даже не знаю, думал об этом, но мне кажется тип ошибки всегда будет тесно связан с типом Чекера.

parshikovalexey replied:
@DronKonstantin, здесь скорее всего для @LaHesis играет роль рекомендация - каждый класс определять в отдельном файле. Это имеет место по той причине, что при разделении классов по отдельным файлам - мы получаем возможность разделять друг от друга разнородного типа объекты. Если MistakeLT - описывает модель, которая представляет сущность данных, то SpellCheckerLT - это уже сервис, который выполняет функциональные операции над данными (если позволите - контроллер в модели MVC). 
С другой стороны, @DronKonstantin скорее всего имеет ввиду, что MistakeLT - необходимая часть для SpellCheckerLT и является его составляющей, поэтому здесь и размещается.
Мое мнение - модели данных нужно отделять и переносить в отдельный файл.

Climber24 commented on BOTFirst/SpellChecker/Classes/SpellCheckerLT.cs:
```diff
@@ -0,0 +1,247 @@
+using System.Linq;
+using System.Text;
+using System.Net;
+using Newtonsoft.Json.Linq;
```
В using Newtonsoft.Json.Linq; в SpellCheckerLT выдает ошибку, что нет Newtonsof, какой пакет нужно установить или какую ссылку добавить, чтобы ошибки не было?

parshikovalexey replied:
Если проект не компилируется из ветки, и выдаются такие ошибки, то значит пакет не установлен. Нужно добавить его в проект, чтобы при развертывании на Азуре - не было проблем.

DronKonstantin replied:
Пакет так и называется Newtonsoft.JSON. Добавить его можно с помощью NuGet. Очень полезная библиотека, сильно упрощает работу с JSON, советую добавить в проект.

Climber24 replied:
Причина была в версии студии, на 2015ой все норм, а на 2017ой выдает, что нет пространства имен.

Climber24 commented on BOTFirst/SpellChecker/Classes/SpellCheckerLT.cs:
```diff
@@ -0,0 +1,247 @@
+        private string getLanguagesUrl; // URL для получения списка доступных языков
+        private string checkResult; // строка(JSON) полученная от LanguageTool API при проверке текста
+        private string languagesresult; // строка(JSON) полученная от LanguageTool API при получении списка доступных языков
+        private Exception error; // дублирующее поле для хранения ошибки полученной при выполнении методов, любое исключение полученное в ходе выполнения метода окажется здесь
```
Может быть стоит переименовать error в checkerError или CheckerError? Т.к. в MistakeLT тоже есть error.

Climber24 commented on BOTFirst/SpellChecker/Classes/SpellCheckerLT.cs:
```diff
@@ -0,0 +1,247 @@
+        /// <para><see cref="CheckAsync(string, string)"/> </para>
+        /// <para><see cref="CorrectMistakes"/> </para>
+        /// </summary>
+        public Exception Error { get { return error; } }
```
Вообще соответствующая приватная переменная свойства (error) по идее должна создаваться автоматически, поэтому ее скорее всего можно просто убрать (свойство будет public Exception Error { get; set; }), т.к. в get кроме возвращения переменной ничего нет. Хотя, если планируется, что туда будет добавлено еще что-то, то это другое дело конечно.

DronKonstantin replied:
https://docs.microsoft.com/ru-ru/dotnet/csharp/programming-guide/classes-and-structs/using-properties
В документации Microsoft рекомендуется создавать приватные переменные для свойств.

Climber24 replied:
Ок, почитаю.

Climber24 commented on BOTFirst/SpellChecker/Classes/SpellCheckerLT.cs:
```diff
@@ -0,0 +1,247 @@
+        /// <summary>
+        /// Создает объект ошибки на основе переданного в конструктор JSON
+        /// </summary>
+        /// <param name="match"></param>
```
Есть ли смысл в пустом теге?

DronKonstantin replied:
Он добавлен автоматически, я забыл убрать

Climber24 commented on BOTFirst/SpellChecker/Classes/SpellCheckerLT.cs:
```diff
@@ -0,0 +1,247 @@
+            int length = Utilities.JSON.GetValueByKey<int>(context, "length", ref error); 
+            string text = Utilities.JSON.GetValueByKey<string>(context, "text", ref error);
+            string short_message = Utilities.JSON.GetValueByKey<string>(match, "shortMessage", ref error);
+            Position = new Position(offset, length);
```
Почему эта переменная с большой буквы, а переменные выше - с маленькой, ведь все они, если не ошибаюсь, приватные?

parshikovalexey replied:
@LaHesis, Position - публичная в родительском классе

Climber24 commented on BOTFirst/SpellChecker/Classes/SpellCheckerLT.cs:
```diff
@@ -0,0 +1,247 @@
+            string short_message = Utilities.JSON.GetValueByKey<string>(match, "shortMessage", ref error);
+            Position = new Position(offset, length);
+            Replacements = replacements.Values<string>("value").ToList();
+            Original = !String.IsNullOrEmpty(text) ? text.Substring(offset, length) : "";
```
Мое мнение - отрицание немного усложняет понимание, может быть лучше убрать его, а ветви тернарного оператора поменять местами?

DronKonstantin replied:
Без проблем

Climber24 commented on BOTFirst/SpellChecker/Classes/SpellCheckerLT.cs:
```diff
@@ -0,0 +1,247 @@
+    public class SpellCheckerLT : ISpellChecker
+    {
+        #region Приватные поля
+        private List<Mistake> mistakes; // дублирующее поле содержащее список ошибок в проверенном тексте
```
Согласно code convention, все комментарии должны начинаться с большой буквы, заканчиваться точкой и располагаться на отдельной строке.

parshikovalexey replied:
+1. Комментарии, которые на отдельной строке и выше кода, к которому относятся - читаются намного лучше.

Climber24 commented on BOTFirst/SpellChecker/Classes/SpellCheckerLT.cs:
```diff
@@ -0,0 +1,247 @@
+        private Exception error; // дублирующее поле для хранения ошибки полученной при выполнении методов, любое исключение полученное в ходе выполнения метода окажется здесь
+        #endregion
+        #region Приватные методы
+        private void GetAvailableLanguagesHandler(object sender, DownloadStringCompletedEventArgs e) // данный метод вызывается при срабатывании события асинхронной загрузки строки(DownloadStringCompleted)
```
Полагаю, что лучше вынести комментарий в отдельную строку перед объявлением метода либо в документационный комментарий для метода + добавить точку и начинать комментарий с большой буквы, согласно code conventions.

parshikovalexey replied:
Для комментирования методов - использовать следует автоматическое документирование через ///. см тут: https://habrahabr.ru/post/41514/. Тут ты сразу и про метод опишешь и параметры также опишешь

Climber24 commented on BOTFirst/SpellChecker/Classes/SpellCheckerLT.cs:
```diff
@@ -0,0 +1,247 @@
+
+        }
+
+        private void ParseCheck() // метод предназначен для парсинга реультата проверки текста
```
Возможно, что стоит переименовать ParseCheck() в ParseCheckResult() или ParseCheckingResult().

parshikovalexey replied:
ParseCheckResult() - лучше

Climber24 commented on BOTFirst/SpellChecker/Classes/SpellCheckerLT.cs:
```diff
@@ -0,0 +1,247 @@
+                {
+                    foreach (var match in matches)
+                    {
+                        MistakeLT curMistake = new MistakeLT(match);
```
Если не ошибаюсь, рекомендуется не использовать сокращения, согласно code conventions, т.е. currentMistake. Хотя у меня в моем коде тоже есть несколько, типа Dic, но там имена длиннее и я на тот момент был не в курсе этого требования.

DronKonstantin replied:
Пусть будет currentMistake

Climber24 commented on BOTFirst/SpellChecker/Classes/SpellCheckerLT.cs:
```diff
@@ -0,0 +1,247 @@
+        /// <param name="language">Код языка для проверки(коды языков можно получить, выполнив методы:
+        /// <see cref="GetAvailableLanguages"/> или <see cref="GetAvailableLanguagesAsync"/>)</param>
+        /// <param name="checkedText">Проверяемый текст</param>
+        public List<Mistake> Check(string language, string checkedText)
```
Предложил бы переименовать checkedText в checkingText или checkableText.

parshikovalexey replied:
вообще можно и просто text. Из имени метода понятно, что делается проверка. Хотя лучше для метода имя будет - CheckText и тогда text -будет ок.

DronKonstantin replied:
На мой взгляд checkedText будет корректным названием, так как это страдательный залог в английском, насколько я помню. Совсем корректным было бы textToBeChecked, но это слишком длинное название.

Climber24 replied:
Чтобы бы был страдательный, нужна to be, а просто checked перед существительным будет лишь как прилагательное (https://www.multitran.ru/c/m.exe?CL=1&l1=1&s=checked) и переводиться это будет как "проверенный текст", что не совсем корректно, ведь мы подаем не проверенный, а проверяемый текст. Еще один вариант можно предложить - textForChecking. Либо просто text, как предлагает Алексей.

parshikovalexey replied:
@DronKonstantin, @LaHesis, на самом деле - наша любовь к таким длинным параметрам(сам этим периодически страдаю) - приводит вот к таким дискуссиям. Если говорить с точки зрения английского, то checkedText - действительно "проверенный текст". И поэтому я предлагаю пытаться переходить к стилю, в котором не нужно будет делать таких имен, а именно - имя метода обозначает конкретное действие, а параметры описывают элементы, над которыми будет проводиться работа. Посмотрите на https://msdn.microsoft.com/ru-ru/library/1k20k614(v=vs.110).aspx и https://msdn.microsoft.com/ru-ru/library/essfb559(v=vs.110).aspx. Из имени метода понятно, что будет делаться - конвертироваться туда-то, или парситься туда-то. И параметры здесь без дополнительной информации - ParsedValue - нет, как нет и ValueToBeConverted. Просто значение. Поэтому и предлагаю вариант с именованием CheckText - проверить текст, а параметр просто text.

Climber24 commented on BOTFirst/SpellChecker/Classes/SpellCheckerLT.cs:
```diff
@@ -0,0 +1,247 @@
+        }
+        public string CorrectMistakes()
+        {
+            return ""; // не реализован
```
Предлагаю использовать здесь в комментарии TODO или NOTE.

Climber24 commented on BOTFirst/SpellChecker/Program.cs:
```diff
@@ -0,0 +1,92 @@
+        static void DemonstrateAsync() // метод для демонстрации асинхронных методов
+        {
+            spellchkr = new SpellCheckerLT();
+            spellchkr.GetAvailableLanguagesAsync(); // получаем список языков(вернее только посылаем запрос)
```
Возможно, что сначала лучше подписаться, а потом сделать запрос?

DronKonstantin replied:
Пожалуй, что так.

Climber24 commented on BOTFirst/SpellChecker/Program.cs:
```diff
@@ -0,0 +1,92 @@
+            var languages = spellchkr.GetAvailableLanguages();
+            var mistakes = spellchkr.Check(languages["Russian"], text);
+            Console.WriteLine(text);
+            Console.WriteLine("------------Ошибки---------------");
```
Предложил бы вынести этот код вывода ошибок, который повторяется или очень похож на код в случае асинхронной демонстрации, в отдельный метод.

parshikovalexey replied:
+1 для хорошего стиля программирования.

Climber24 commented on BOTFirst/SpellChecker/Program.cs:
```diff
@@ -0,0 +1,92 @@
+        static void MistakesDownloaded() // обработчик события получения результатов проверки
+        {
+            if (spellchkr.Error == null)
+            {
```
Не должен ли здесь быть вывод исходного текста Console.WriteLine(text);, как в случае синхронной проверки? Либо, если вынести этот вывод результатов в отдельный метод, то вызов этого отдельного метода вместо несколько повторяющегося кода.

DronKonstantin replied:
Я делал методы демонстрации на скорую руку, только чтобы продемонстрировать работу класса. Все верно, стоит вынести вывод в отдельный метод.

Climber24 reviewed (commented):
В целом отличная работа, каких-либо серьезных ошибок я не нашел, мои комментарии в основном на тему code convention'а. Возможно, если бы у меня был опыт использования асинхронных методов, то нашел бы что-то, что стоит поменять. PR я бы одобрил, но пусть лучше посмотрит Алексей, поэтому просто использую Comment.

parshikovalexey commented on BOTFirst/SpellChecker/Classes/SpellCheckerLT.cs:
```diff
@@ -0,0 +1,247 @@
+            languages = new Dictionary<string, string>();
+            try
+            {
+                JToken parsedJSON = JToken.Parse(languagesresult);
```
@DronKonstantin, я уже Сергею писал. Опять таки по code convention - если из правой части выражения понятно - какого типа должна быть переменная, то в c# рекомендуется использовать var вместо типа переменной.

DronKonstantin replied:
JToken parsedJSON = JToken.Parse(languagesresult);
На мой взгляд тип переменной не слишком понятен из наименования метода.

parshikovalexey replied:
@DronKonstantin, ок тогда. Мне было понятно какой тип, поэтому и привел комментарий здесь

parshikovalexey committed:
1. Обновил json библиотеки во всем проекте к единой версии
2. Сделал по другому асинхронные методы и убрал переменные контекста из интерфейса.
Дело в том, что как заметил Сергей, данный интерфейс получился больше похожим на абстрактный класс. Я его немного переделал (не все, еще нужно обработать варианты с ошибками, т.к. Errors оставил). Интерфес по сути должен содержать только методы, которые нужно выполнять, и не содержать дополнительные свойства. Плюс я не очень хорошо отношусь к классам, в которых реализация построена таким образом. Метод А вычисляет что-то, присваиает значение переменной Х, вызывает метод Б(который вычилсяет что-то из Х и присвает это в У), а потом возвращает У. Это приводит к тому, что в классе появляются приватные переменные, которые нужно присваивать значениям, чтобы другие методы с ними могли работать. А вот это несет эффект того, что где-то кто-то(в смысле другой метод) может обновить переменную Х, а наш исходный метод будет не знать об этом. Это в свою очередь приводит к side-эффектам различного рода и может вызывать скрытые проблемы. ЛУчше не хранить этот контекст, а использовать чистые методы, которые не знают о контексте. В этом случае проще найти ошибку и сложнее ее допустить и при доработке - не нужно проверять кучу другого кода, чтобы не накосячить чужоей контекст. То есть мы реализуем поведение, когда метод А вычисляет в локальную Х, а потом вызывает метод У с параметром Х, который возвращает локально расчитанный У и мы с ним можем уже оперировать и не бояться что-то кто-то еще нашу переменную У в это время поменяет.
Посмотрите как я для этого переделал методы интерфейса.
И еще, ребята, в моем коде оформление { по другому. У меня студия настроена на не перенос открывающей { на новую строку. То есть со стилем Вашего кода расходится.

parshikovalexey reviewed (changes requested):
@DronKonstantin, Костя. Я там набросал комментариев, которые нужно посмотреть. Плюс я там реализацию класса на более чистые методы (свободные от контекста). Хотел, чтобы ты посмотрел подход и его возможно обсудили.
По запросу на изменения: давай комментарии перенесем сверху, как @LaHesis  подметил, и у переменных сделаем var вместо типов. И вдогонку можно отшлифовать те вещи, которые @LaHesis упомянул

LaHesis commented on BOTFirst/SpellChecker/Classes/SpellCheckerLT.cs:
```diff
@@ -0,0 +1,247 @@
+        /// <para><see cref="CheckAsync(string, string)"/> </para>
+        /// <para><see cref="CorrectMistakes"/> </para>
+        /// </summary>
+        public Exception Error { get { return error; } }
```
Ок

DronKonstantin committed:
Последние штрихи

Исправил указанные мне ошибки
1. Поставил var вместо типа у переменных
2. Перенес комментарии описывающие методы в блок summary

Climber24 commented on BOTFirst/SpellChecker/Classes/SpellCheckerLT.cs:
```diff
@@ -0,0 +1,205 @@
+    public class SpellCheckerLT : ISpellChecker
+    {
+        #region Приватные поля
+        private readonly string checkUrl; // URL для проверки текста
```
Здесь и еще в нескольких местах - согласно code conventions, комментарии должны располагаться на отдельной строке, начинаться с большой буквы и заканчиваться точкой. Поправь, если не трудно.

DronKonstantin replied:
Исправлено

DronKonstantin committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into SpellChecker

# Conflicts:
#	BOTFirst/AchievementBot.sln

DronKonstantin committed:
Исправления по Code Convention

Эти изменения не несут функциональности, но приводят код к общему виду.

parshikovalexey committed:
Убрал остатки конфликтов после мержа с веткой develop
